### PR TITLE
(678) Allow sorting on dependent content

### PR DESCRIPTION
**Pull request by pezholio**
This allows sorting and pagination on dependent content, both on the view and review pages

## Screenshots

_Note: I temporarily changed the number of results per page to demonstrate the pagination_

image

image

### CODE REVIEW

File: lib/engines/content_block_manager/app/services/content_block_manager/get_host_content_items.rb [hunk 1]
```diff
@@ -4,18 +4,22 @@
...
+    def initialize(content_id:, page: nil, order: nil)
       self.content_id = content_id
+      self.page = page
+      self.order = order || DEFAULT_ORDER
     end
 
-    def self.by_embedded_document(content_block_document:)
-      new(content_id: content_block_document.content_id).items
+    def self.by_embedded_document(content_block_document:, page: nil, order: nil)
+      new(content_id: content_block_document.content_id, page:, order:).items
```
Comment (Harriethw) [inline lib/engines/content_block_manager/app/services/content_block_manager/get_host_content_items.rb:18]: I wondered if you'd seen the `Kaminara` helper methods that Whitehall uses to paginate? https://github.com/alphagov/whitehall/blob/c88f75b785b92b5d045ac1e2d35c675ed98c414c/app/models/admin/edition_filter.rb#L26 
That means the results can just be passed to the `paginate` function in the `erb` file https://github.com/alphagov/whitehall/blob/adee19eb5cff5c1c83cb0b2ab46ece334539bd0b/lib/engines/content_block_manager/app/views/content_block_manager/content_block/documents/index.html.erb#L40 
it kind of hides away the logic/shape of data but is a bit cleaner, and you wouldn't need the extra data class that's been created?

File: lib/engines/content_block_manager/app/services/content_block_manager/get_host_content_items.rb [hunk 2]
```diff
@@ -4,18 +4,22 @@
...
   class GetHostContentItems
-    attr_reader :content_id
+    attr_reader :content_id, :page, :order
 
-    def initialize(content_id:)
+    DEFAULT_ORDER = "-unique_pageviews".freeze
+
+    def initialize(content_id:, page: nil, order: nil)
       self.content_id = content_id
+      self.page = page
```
Comment (Harriethw) [inline lib/engines/content_block_manager/app/services/content_block_manager/get_host_content_items.rb:13]: minor question: is having the `self` attributes here a more old fashioned way of having class attributes, over the `@` ones?

File: lib/engines/content_block_manager/app/services/content_block_manager/get_host_content_items.rb [hunk 3]
```diff
@@ -4,18 +4,22 @@
 
 module ContentBlockManager
   class GetHostContentItems
-    attr_reader :content_id
+    attr_reader :content_id, :page, :order
 
-    def initialize(content_id:)
+    DEFAULT_ORDER = "-unique_pageviews".freeze
```
Comment (Harriethw) [inline lib/engines/content_block_manager/app/services/content_block_manager/get_host_content_items.rb:9]: does this mean it shows the pages with least page views first? Is this something we'll revisit, as I guess the more page views the more important the document is?

**Review by Harriethw — APPROVED** [review]
Looks good! just had a couple of questions more for clarification really

File: lib/engines/content_block_manager/app/services/content_block_manager/get_host_content_items.rb (hunk 1 above)
Comment (pezholio) [reply lib/engines/content_block_manager/app/services/content_block_manager/get_host_content_items.rb:18]: Yeah, we'd have to load all the results into memory then wouldn't we? (although, on reflection, that might have been easier in the first instance 😅 )

File: lib/engines/content_block_manager/app/services/content_block_manager/get_host_content_items.rb (hunk 2 above)
Comment (pezholio) [reply lib/engines/content_block_manager/app/services/content_block_manager/get_host_content_items.rb:13]: It's more akin to having an `@` variable with an `attr_reader`, which I kinda prefer and is more Rubyish to my eyes. Think I'll get into the habit of removing this pattern when I see it.

File: lib/engines/content_block_manager/app/services/content_block_manager/get_host_content_items.rb (hunk 3 above)
Comment (pezholio) [reply lib/engines/content_block_manager/app/services/content_block_manager/get_host_content_items.rb:9]: This confused me too, but when we're doing ascending order with numbers, it shows the smallest first, so to see the most page views first we need descending order.

**Pull request merged**